app/helpers/fhir_helper.rb

```diff
@@ -232,6 +232,67 @@ def close_contact_from_fhir(related_person)
     }
   end
 
+  # Returns a representative FHIR::Immunization for an instance of a Sara Alert Vaccine.
+  # https://www.hl7.org/fhir/immunization.html
+  def vaccine_as_fhir(vaccine)
+    FHIR::Immunization.new(
+      meta: FHIR::Meta.new(lastUpdated: vaccine.updated_at.strftime('%FT%T%:z')),
+      id: vaccine.id,
+      status: 'completed',
+      vaccineCode: [
+        FHIR::CodeableConcept.new(
+          text: vaccine.product_name,
+          coding: Vaccine.product_codes_by_name(vaccine.group_name, vaccine.product_name).map do |code|
+            FHIR::Coding.new(code: code['code'], system: code['system'])
+          end
+        )
+      ],
+      patient: FHIR::Reference.new(reference: "Patient/#{vaccine.patient_id}"),
+      occurrenceDateTime: vaccine.administration_date,
+      note: FHIR::Annotation.new(text: vaccine.notes),
+      protocolApplied: [
+        {
+          targetDisease: [
+            FHIR::CodeableConcept.new(
+              text: vaccine.group_name,
+              coding: Vaccine.group_codes_by_name(vaccine.group_name).map do |code|
+                FHIR::Coding.new(code: code['code'], system: code['system'])
+              end
+            )
+          ],
+          doseNumberString: vaccine.dose_number
+        }
+      ]
+    )
+  end
+
+  # Create a hash of atttributes that corresponds to a Sara Alert Patient (and can be used to
+  # create new ones, or update existing ones), using the given FHIR::Patient.
+  # Hash is of the form:
+  # {
+  #  attribute_name: { value: <converted-value>, path: <fhirpath-to-corresponding-fhir-element>, errors: <array-of-messages> }
+  # }
+  def vaccine_from_fhir(vaccine)
+    group_coding = vaccine&.protocolApplied&.first&.targetDisease&.first&.coding&.first
+    group = Vaccine.group_name_by_code(group_coding&.system, group_coding&.code)
+    group_errors = ["is not an acceptable value, acceptable values are: #{Vaccine.group_code_options.join(', ')}"] if group.nil? && !group_coding.nil?
+
+    product_name_coding = vaccine&.vaccineCode&.first&.coding&.first
+    product_name = Vaccine.product_name_by_code(group, product_name_coding&.system, product_name_coding&.code)
+    if product_name.nil? && !product_name_coding.nil?
+      product_name_errors = ["is not an acceptable value, acceptable values are: #{Vaccine.product_code_options(group).join(', ')}"]
+    end
+
+    {
+      group_name: { value: group, path: 'Immunization.protocolApplied[0].targetDisease[0].coding[0]', errors: group_errors },
+      product_name: { value: product_name, path: 'Immunization.vaccineCode[0].coding[0]', errors: product_name_errors },
```

Comment:
I've kind of gone back and forth on this, but how would you feel about adding a `.code` on the end of these paths so that we just point to the code value. The quote-escaped JSON object in the error message is a little hard to parse visually.

Reply:
Yeah I agree the JSON objects are not ideal to look at, but to just show the `code` is a bit misleading, since really it is the `system`, `code` pair that defines if the value can be used or not. I'll look at it and see if I can find a better way to write out the JSON object so it isn't as confusing to look at. Does that work?

Reply:
I think if the message is worded appropriately, just pointing at the code is probably fine, but if it's not too much effort to format it better and include the system, then even better

Reply:
I updated this in 9ed52ca to use a syntax like `<system>#<code>`. So an error message will look like this:
```
"Value 'http://terminology.hl7.org/CodeSystem/v3-NullFlavor#UNKs' for 'Product Name' is not an acceptable value, acceptable values are: http://hl7.org/fhir/sid/cvx#207, http://hl7.org/fhir/sid/cvx#208, http://hl7.org/fhir/sid/cvx#212, http://terminology.hl7.org/CodeSystem/v3-NullFlavor#UNK"
```
To me this is pretty readable, and I think the syntax gets the idea across. But, this isn't a standard syntax or anything, so let me know if you don't like it. Maybe it would be better to just do something like `{system: <system>, code: <code>}` without any `\"` characters. But it is pretty important to me that the `system` is included, because I think someone would be pretty confused wondering why their code is being rejected if it is from the list of valid codes, but they have an invalid system.

Reply:
I'm good with this. I think it's easy enough to determine what the syntax means.